### ISSUE TITLE
[ENH]: Improving Parallelization Efficiency

### ISSUE THREAD

**Issue by celestinoxp** [pycaret/pycaret#3625]
### pycaret version checks

- [X] I have checked that this issue has not already been reported [here](https://github.com/pycaret/pycaret/issues).

- [X] I have confirmed this bug exists on the [latest version](https://github.com/pycaret/pycaret/releases) of pycaret.

- [X] I have confirmed this bug exists on the master branch of pycaret (pip install -U git+https://github.com/pycaret/pycaret.git@master).

### Issue Description

I can't get parallelization to work efficiently on my dataset. I've done everything and the cpu doesn't use more than 35% of the cpu, and only 1 core is used at 100%, the rest have very low usage.
The goal is to use all cpu cores at or near 100%.

I converted all categorical values to numbers through mapping and I converted all nans to zero. the problem continues.

Therefore, I passed the dataset to a csv and created a reproducible code. Therefore, I hope it helps to develop pycaret for better parallelization efficiency.

Dataframe to reproduce:
[meu_dataframe.csv](https://github.com/pycaret/pycaret/files/11881429/meu_dataframe.csv)

### Reproducible Example

```python
import pandas as pd
from pycaret.classification import *

df = pd.read_csv('meu_dataframe.csv')

df.head()

setup(data=df, target = 'soma_menor_que_47', fold=5, fold_shuffle=False, polynomial_features=False, ignore_features=['data', 'n1', 'n2', 'n3', 'n4', 'n5', 'ns', 'soma', 'posicao_chave', 'primos5', 'soma_n1_n2', 'soma_n1_n3', 'soma_n1_n4', 'soma_n1_n5', 'soma_n2_n3', 'soma_n2_n4', 'soma_n2_n5', 'soma_n3_n4', 'soma_n3_n5', 'soma_n4_n5','linha1', 'linha2', 'linha3', 'linha4', 'linha5', 'linha6', 'linha7', 'coluna1', 'coluna2', 'coluna3', 'coluna4', 'coluna5', 'coluna6', 'coluna7', 'cor_n1', 'cor_n2', 'cor_n3', 'cor_n4', 'cor_n5', 'n1_p_i', 'n2_p_i', 'n3_p_i', 'n4_p_i', 'n5_p_i', 'anel_1', 'anel_2', 'anel_3', 'qx1', 'qx2', 'qx3', 'qx4',
           'qq1', 'qq2', 'qq3', 'qq4', 'q1', 'q2', 'q3', 'q4', 'qp1', 'qp2', 'qp3', 'qp4', 'qp5', 'fppc_n1', 'fppc_n2', 'fppc_n3', 'fppc_n4', 'fppc_n5',
           'n1_primo', 'n2_primo', 'n3_primo', 'n4_primo', 'n5_primo', 'n1_sdi', 'n2_sdi', 'n3_sdi', 'n4_sdi', 'n5_sdi', 'n_1_10', 'n_11_20', 'n_21_30', 'n_31_40',
           'n_41_49', 'num_sorteios_last_premio', 'n1_ausencias', 'n2_ausencias', 'n3_ausencias', 'n4_ausencias', 'n5_ausencias', 'n1_media_last5', 
           'n2_media_last5', 'n3_media_last5', 'n4_media_last5', 'n5_media_last5', 'n1_media_last10', 'n2_media_last10', 'n3_media_last10', 'n4_media_last10',
           'n5_media_last10', 'n1_n2_ausencias', 'n1_n3_ausencias', 'n1_n4_ausencias', 'n1_n5_ausencias', 'n2_n3_ausencias', 'n2_n4_ausencias', 'n2_n5_ausencias',
           'n3_n4_ausencias', 'n3_n5_ausencias', 'n4_n5_ausencias'], fold_strategy='timeseries', use_gpu=True, normalize=True, session_id=123)

xgboost = create_model('xgboost')

tuned_xgboost = tune_model(xgboost, n_iter=50)
```

### Expected Behavior

cpu with all cores in use at 100% or close to it. or the gpu at 100% or close to it.

### Actual Results

```python-traceback
only 1 core at 100%
```

### Installed Versions

<details>
PyCaret required dependencies:
                 pip: 23.1.2
          setuptools: 68.0.0
             pycaret: 3.0.3
             IPython: 8.14.0
          ipywidgets: 8.0.6
                tqdm: 4.65.0
               numpy: 1.25.0
              pandas: 2.0.2
              jinja2: 3.1.2
               scipy: 1.11.0
              joblib: 1.3.0.dev0
             sklearn: 1.2.2
                pyod: 1.1.0
            imblearn: 0.10.1
   category_encoders: 2.6.1
            lightgbm: 3.3.5
               numba: 0.57.1
            requests: 2.31.0
          matplotlib: 3.7.1
          scikitplot: 0.3.7
         yellowbrick: 1.5
              plotly: 5.15.0
    plotly-resampler: Not installed
             kaleido: 0.2.1
           schemdraw: 0.17
         statsmodels: 0.14.0
              sktime: 0.20.0
               tbats: 1.1.3
            pmdarima: 2.0.3
              psutil: 5.9.5
          markupsafe: 2.1.3
             pickle5: Not installed
         cloudpickle: 2.2.1
         deprecation: 2.1.0
              xxhash: 3.2.0
           wurlitzer: Not installed

PyCaret optional dependencies:
                shap: 0.41.0
           interpret: 0.4.2
                umap: 0.5.3
    pandas_profiling: 4.3.1
  explainerdashboard: 0.4.2.2
             autoviz: 0.1.720
           fairlearn: 0.8.0
          deepchecks: 0.17.3
             xgboost: 2.0.0-dev
            catboost: 1.2
              kmodes: 0.12.2
             mlxtend: 0.22.0
       statsforecast: 1.5.0
        tune_sklearn: 0.4.5
                 ray: 2.5.1
            hyperopt: 0.2.7
              optuna: 3.2.0
               skopt: 0.9.0
              mlflow: 2.4.1
              gradio: 3.35.2
             fastapi: 0.98.0
             uvicorn: 0.22.0
              m2cgen: 0.10.0
           evidently: 0.3.3
               fugue: 0.8.5
           streamlit: Not installed
             prophet: Not installed
</details>

**Comment by trivialfis** [comment]
Hi, got an email from @celestinoxp . I looked into the dataset and it's quite small, XGBoost can finish training a hundred trees using `XGBClassifier` within 0.285 seconds. For such a short time period, it's difficult for simple sampling monitors to get an accurate estimation of CPU usage. I suggest that you either use a real profiler to see the CPU usage of training or find a larger dataset for testing purposes (like replicating the dataset a few thousand times).

Aside from this, many Python libraries like numpy and pandas use a single core for their algorithms, if these libraries take up the majority of runtime, you might not see the effect of XGBoost using multiple threads from simple monitors like htop.

**Comment by celestinoxp** [comment]
more context: problem is about create_model and tune_model and xgboost algorithm

**Comment by celestinoxp** [comment]
what is means?
2023-07-04 23:20:30,627:WARNING:C:\Users\username\anaconda3\lib\site-packages\pycaret\internal\pipeline.py:260: UserWarning: Persisting input arguments took 1.07s to run.If this happens often in your code, it can cause performance problems (results will be correct in all cases). The reason for this is probably some large input arguments for a wrapped function.
  fitted_estimator = self._memory_fit(

2023-07-04 23:20:37,337:WARNING:C:\Users\username\anaconda3\lib\site-packages\pycaret\internal\pipeline.py:310: UserWarning: Persisting input arguments took 2.65s to run.If this happens often in your code, it can cause performance problems (results will be correct in all cases). The reason for this is probably some large input arguments for a wrapped function.
  X, _ = self._memory_full_transform(self, X, None, with_final=False)

**Comment by celestinoxp** [comment]
Hello again,
yesterday I tested naive bayes and in tune_model, used all cores/threads at 100%, but all metrics were equal to create_model.

If so, then could pycaret be using the wrong xgboost configuration?

@ngupta23 @trivialfis @tvdboom @Yard1 can investigate?

**Comment by celestinoxp** [comment]
logs.log:
2023-07-10 11:40:14,047:WARNING:C:\Users\username\anaconda3\lib\site-packages\pycaret\internal\pipeline.py:310: UserWarning: Persisting input arguments took 3.88s to run.If this happens often in your code, it can cause performance problems (results will be correct in all cases). The reason for this is probably some large input arguments for a wrapped function.
  X, _ = self._memory_full_transform(self, X, None, with_final=False)

2023-07-10 11:40:27,991:WARNING:C:\Users\username\anaconda3\lib\site-packages\pycaret\internal\pipeline.py:260: UserWarning: Persisting input arguments took 1.72s to run.If this happens often in your code, it can cause performance problems (results will be correct in all cases). The reason for this is probably some large input arguments for a wrapped function.
  fitted_estimator = self._memory_fit(

2023-07-10 11:40:37,306:WARNING:C:\Users\username\anaconda3\lib\site-packages\pycaret\internal\pipeline.py:310: UserWarning: Persisting input arguments took 4.42s to run.If this happens often in your code, it can cause performance problems (results will be correct in all cases). The reason for this is probably some large input arguments for a wrapped function.
  X, _ = self._memory_full_transform(self, X, None, with_final=False)

2023-07-10 11:40:50,522:WARNING:C:\Users\username\anaconda3\lib\site-packages\pycaret\internal\pipeline.py:260: UserWarning: Persisting input arguments took 1.69s to run.If this happens often in your code, it can cause performance problems (results will be correct in all cases). The reason for this is probably some large input arguments for a wrapped function.
  fitted_estimator = self._memory_fit(

**Comment by celestinoxp** [comment]
I found a way where we can analyze everything create_model('xgboost') does line-by-line and we can analyze everything we can optimize in addition to the third-party packages we can invite to optimize their code.

Reproduzible code:

import cProfile
import pstats

import pandas as pd
from pycaret.classification import *

df = pd.read_csv('meu_dataframe.csv')

# Função a ser perfilada
def my_create_model(*args, **kwargs):
    return create_model(*args, **kwargs)
    
    setup(data=df, target = 'soma_menor_que_47', fold=5, fold_shuffle=False, polynomial_features=False, ignore_features=['data', 'n1', 'n2', 'n3', 'n4', 'n5', 'ns', 'soma', 'posicao_chave', 'primos5', 'soma_n1_n2', 'soma_n1_n3', 'soma_n1_n4', 'soma_n1_n5', 'soma_n2_n3', 'soma_n2_n4', 'soma_n2_n5', 'soma_n3_n4', 'soma_n3_n5', 'soma_n4_n5','linha1', 'linha2', 'linha3', 'linha4', 'linha5', 'linha6', 'linha7', 'coluna1', 'coluna2', 'coluna3', 'coluna4', 'coluna5', 'coluna6', 'coluna7', 'cor_n1', 'cor_n2', 'cor_n3', 'cor_n4', 'cor_n5', 'n1_p_i', 'n2_p_i', 'n3_p_i', 'n4_p_i', 'n5_p_i', 'anel_1', 'anel_2', 'anel_3', 'qx1', 'qx2', 'qx3', 'qx4',
           'qq1', 'qq2', 'qq3', 'qq4', 'q1', 'q2', 'q3', 'q4', 'qp1', 'qp2', 'qp3', 'qp4', 'qp5', 'fppc_n1', 'fppc_n2', 'fppc_n3', 'fppc_n4', 'fppc_n5',
           'n1_primo', 'n2_primo', 'n3_primo', 'n4_primo', 'n5_primo', 'n1_sdi', 'n2_sdi', 'n3_sdi', 'n4_sdi', 'n5_sdi', 'n_1_10', 'n_11_20', 'n_21_30', 'n_31_40',
           'n_41_49', 'num_sorteios_last_premio', 'n1_ausencias', 'n2_ausencias', 'n3_ausencias', 'n4_ausencias', 'n5_ausencias', 'n1_media_last5', 
           'n2_media_last5', 'n3_media_last5', 'n4_media_last5', 'n5_media_last5', 'n1_media_last10', 'n2_media_last10', 'n3_media_last10', 'n4_media_last10',
           'n5_media_last10', 'n1_n2_ausencias', 'n1_n3_ausencias', 'n1_n4_ausencias', 'n1_n5_ausencias', 'n2_n3_ausencias', 'n2_n4_ausencias', 'n2_n5_ausencias',
           'n3_n4_ausencias', 'n3_n5_ausencias', 'n4_n5_ausencias'], fold_strategy='timeseries', use_gpu=True, normalize=True, session_id=123)
           
# Executa o código com o cProfile
pr = cProfile.Profile()
pr.enable()

xgboost = my_create_model('xgboost')

pr.disable()
ps = pstats.Stats(pr)
ps.print_stats()

**Comment by moezali1** [comment]
Closing as part of the 4.0 backlog cleanup — see #4177.

PyCaret 4.0 is a ground-up rewrite (OOP-only API, sklearn-native engine, Python 3.11+, sklearn ≥ 1.7) and almost all pre-4.0 issues either reference removed APIs (`setup()`, `plot_model`, `evaluate_model`, `interpret_model`, `automl`, the legacy time-series module, the Streamlit dashboard, …) or no longer reproduce on the new code path.

**If this still affects you in 4.0, please re-file as a new issue** with:

- the experiment constructor call (`Experiment(...)`),
- the operation that fails,
- the full traceback,
- your Python / sklearn / pycaret versions.

Resources for 4.0:

- Docs: https://pycaret.org/docs
- Migrate from 3.x: https://pycaret.org/docs/guides/migrating-from-3-x
- API reference: https://pycaret.org/reference
- FAQ: https://pycaret.org/docs/resources/faqs

Thanks for filing — and apologies for the long latency. The repo had drifted; we're starting clean.
